components/zcash_note_encryption/src/batch.rs

```diff
@@ -0,0 +1,69 @@
+//! APIs for batch trial decryption.
+
+use std::iter;
+
+use crate::{
+    try_compact_note_decryption_inner, try_note_decryption_inner, Domain, EphemeralKeyBytes,
+    ShieldedOutput,
+};
+
+/// Trial decryption of a batch of notes with a set of recipients.
+///
+/// This is the batched version of [`zcash_note_encryption::try_note_decryption`].
+pub fn try_note_decryption<D: Domain, Output: ShieldedOutput<D>>(
+    ivks: &[D::IncomingViewingKey],
+    outputs: &[(D, Output)],
+) -> Vec<Option<(D::Note, D::Recipient, D::Memo)>> {
+    batch_note_decryption(ivks, outputs, try_note_decryption_inner)
+}
+
+/// Trial decryption of a batch of notes for light clients with a set of recipients.
+///
+/// This is the batched version of [`zcash_note_encryption::try_compact_note_decryption`].
+pub fn try_compact_note_decryption<D: Domain, Output: ShieldedOutput<D>>(
+    ivks: &[D::IncomingViewingKey],
+    outputs: &[(D, Output)],
+) -> Vec<Option<(D::Note, D::Recipient)>> {
+    batch_note_decryption(ivks, outputs, try_compact_note_decryption_inner)
+}
+
+fn batch_note_decryption<D: Domain, Output: ShieldedOutput<D>, F, FR>(
+    ivks: &[D::IncomingViewingKey],
+    outputs: &[(D, Output)],
+    decrypt_inner: F,
+) -> Vec<Option<FR>>
+where
+    F: Fn(&D, &D::IncomingViewingKey, &EphemeralKeyBytes, &Output, D::SymmetricKey) -> Option<FR>,
+{
+    // Fetch the ephemeral keys for each output and batch-parse them.
+    let ephemeral_keys = D::batch_epk(outputs.iter().map(|(_, output)| output.ephemeral_key()));
+
+    // Derive the shared secrets for all combinations of (ivk, output).
+    // The scalar multiplications cannot benefit from batching.
```

Comment:
They could benefit from parallelisation, maybe. But rayon or crossbeam is not currently a dependency.

Reply:
Yes, I expect any of that parallelism to be managed by the caller outside these APIs.